skyportal/tests/api/test_annotations_on_spectrum.py

```diff
@@ -0,0 +1,471 @@
+
+    # try posting without an origin...
+    status, data = api(
+        'POST',
+        f'spectra/{spectrum_id}/annotations',
+        data={
+            'spectrum_id': spectrum_id,
+            'data': {'offset_from_host_galaxy': 1.5},
+            'group_ids': [public_group.id],
+        },
+        token=annotation_token,
+    )
+    assert status == 400
+    assert 'expected string or bytes-like object' in data["message"]
+
+    # this should not work, since "origin" is empty
+    status, data = api(
+        'POST',
+        f'spectra/{spectrum_id}/annotations',
+        data={
+            'origin': '',
+            'data': {'offset_from_host_galaxy': 1.5},
+            'group_ids': [public_group.id],
+        },
+        token=annotation_token,
+    )
+
+    assert status == 400
+    assert 'Input `origin` must begin with alphanumeric/underscore' in data["message"]
+
+    # first time adding an annotation to this object from Kowalski
+    status, data = api(
+        'POST',
+        f'spectra/{spectrum_id}/annotations',
+        data={
+            'origin': 'kowalski',
+            'data': {'offset_from_host_galaxy': 1.5},
+            'group_ids': [public_group.id],
+        },
+        token=annotation_token,
+    )
+
+    assert status == 200
+    annotation_id = data['data']['annotation_id']
+
+    # this should not work, since "origin" Kowalski was already posted to this object
+    # instead, try updating the existing annotation if you have new information!
+    status, data = api(
+        'POST',
+        f'spectra/{spectrum_id}/annotations',
+        data={
+            'origin': 'kowalski',
+            'data': {'offset_from_host_galaxy': 1.5},
+            'group_ids': [public_group.id],
+        },
+        token=annotation_token,
+    )
+
+    assert status == 400
+    assert 'duplicate key value violates unique constraint' in data["message"]
+
+    status, data = api(
+        'GET',
+        f'spectra/{spectrum_id}/annotations/{annotation_id}',
+        token=annotation_token,
+    )
+
+    assert status == 200
+    assert data['data']['data']['offset_from_host_galaxy'] == 1.5
+
+
+def test_add_and_retrieve_comment_group_access(
+    comment_token_two_groups,
+    upload_data_token_two_groups,
+    public_source_two_groups,
+    public_group2,
+    public_group,
+    comment_token,
+    lris,
+):
+    status, data = api(
+        'POST',
+        'spectrum',
+        data={
+            'obj_id': str(public_source_two_groups.id),
+            'observed_at': str(datetime.datetime.now()),
+            'instrument_id': lris.id,
+            'wavelengths': [664, 665, 666],
+            'fluxes': [234.2, 232.1, 235.3],
+            'group_ids': [public_group2.id],
+        },
+        token=upload_data_token_two_groups,
+    )
+    assert status == 200
+    assert data['status'] == 'success'
+    spectrum_id = data["data"]["id"]
+
+    status, data = api(
+        'POST',
+        f'spectra/{spectrum_id}/comments',
```

Comment:
Isn't this supposed to be for annotations? This is the comments API

Reply:
Oh god how did I miss that...